Fix findByIdAndUpdate behavior with undefined id

gramliu opened the pull request:
**Summary**

findByIdAndUpdate should not work with an undefined id, if an update is specified. Resolves #10964

**Examples**

I added a test to demonstrate this change.

vkarpov15 commented on lib/model.js:
```diff
@@ -2676,6 +2680,9 @@ Model.findByIdAndUpdate = function(id, update, options, callback) {
     }
     return this.findOneAndUpdate({ _id: id }, undefined);
+  } else if (id === undefined) {
+    const msg = 'Model.findByIdAndUpdate(): id cannot be undefined if update is specified.\n';
```
Why `\n`?

gramliu replied:
Oh, I was just following the format of the preceding error messages. I went ahead and removed it

vkarpov15 commented on lib/model.js:
```diff
@@ -2676,6 +2680,9 @@ Model.findByIdAndUpdate = function(id, update, options, callback) {
       throw new TypeError(msg);
     }
     return this.findOneAndUpdate({ _id: id }, undefined);
+  } else if (id === undefined) {
```
Our tests rely on `Model.findByIdAndUpdate()` with no arguments working, so can we make this `else if (arguments.length > 0 && id === undefined)`?

gramliu replied:
Thank you for bringing this up. I forgot to consider that. I went ahead and added this to the condition

vkarpov15 reviewed (changes requested):
I think this PR is helpful, but causes some test failures. We can merge this when those get sorted out

gramliu commented:
@vkarpov15 I pushed some changes that might resolve the failing tests. For the function itself, I added the check for undefined id only if the length of arguments was greater than 1. Moreover, I updated the test I wrote to properly check that the error is thrown. It turns out that my original test was wrong

vkarpov15 reviewed (approved):
Needed to patch up the tests but LGTM overall :+1:

vkarpov15 commented:
Unfortunately I have to revert this change - it is backwards breaking as indicated in #11149. We'll consider this for our next major release.
